File: Kalkulator/app/src/main/java/com/zukron/kalkulator/MainActivity.java
package com.zukron.kalkulator;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private TextView tvHistory, tvCalculateAndResult;
    private Button btnNumber0, btnNumber1, btnNumber2, btnNumber3, btnNumber4, btnNumber5, btnNumber6, btnNumber7, btnNumber8, btnNumber9;
    private Button btnCSymbol, btnTimeSymbol, btnDivideSymbol, btnMinusSymbol, btnPlusSymbol, btnEqualSymbol;
    private Button btnBackspace, btnDot;
    private String symbols = "";
    private String text = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvHistory = findViewById(R.id.tv_history);
        tvCalculateAndResult = findViewById(R.id.tv_calculate_and_result);

        btnNumber0 = findViewById(R.id.btn_number_0);
        btnNumber0.setOnClickListener(this);
        btnNumber1 = findViewById(R.id.btn_number_1);
        btnNumber1.setOnClickListener(this);
        btnNumber2 = findViewById(R.id.btn_number_2);
        btnNumber2.setOnClickListener(this);
        btnNumber3 = findViewById(R.id.btn_number_3);
        btnNumber3.setOnClickListener(this);
        btnNumber4 = findViewById(R.id.btn_number_4);
        btnNumber4.setOnClickListener(this);
        btnNumber5 = findViewById(R.id.btn_number_5);
        btnNumber5.setOnClickListener(this);
        btnNumber6 = findViewById(R.id.btn_number_6);
        btnNumber6.setOnClickListener(this);
        btnNumber7 = findViewById(R.id.btn_number_7);
        btnNumber7.setOnClickListener(this);
        btnNumber8 = findViewById(R.id.btn_number_8);
        btnNumber8.setOnClickListener(this);
        btnNumber9 = findViewById(R.id.btn_number_9);
        btnNumber9.setOnClickListener(this);

        btnCSymbol = findViewById(R.id.btn_c_symbol);
        btnCSymbol.setOnClickListener(this);
        btnTimeSymbol = findViewById(R.id.btn_time_symbol);
        btnTimeSymbol.setOnClickListener(this);
        btnDivideSymbol = findViewById(R.id.btn_divide_symbol);
        btnDivideSymbol.setOnClickListener(this);
        btnMinusSymbol = findViewById(R.id.btn_minus_symbol);
        btnMinusSymbol.setOnClickListener(this);
        btnPlusSymbol = findViewById(R.id.btn_plus_symbol);
        btnPlusSymbol.setOnClickListener(this);

        btnBackspace = findViewById(R.id.btn_backspace);
        btnBackspace.setOnClickListener(this);
        btnDot = findViewById(R.id.btn_dot);
        btnDot.setOnClickListener(this);

        btnEqualSymbol = findViewById(R.id.btn_equal_symbol);
        btnEqualSymbol.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                calculate(tvCalculateAndResult.getText().toString());
            }
        });
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btn_number_0:
                text += "0";
                break;
            case R.id.btn_number_1:
                text += "1";
                break;
            case R.id.btn_number_2:
                text += "2";
                break;
            case R.id.btn_number_3:
                text += "3";
                break;
            case R.id.btn_number_4:
                text += "4";
                break;
            case R.id.btn_number_5:
                text += "5";
                break;
            case R.id.btn_number_6:
                text += "6";
                break;
            case R.id.btn_number_7:
                text += "7";
                break;
            case R.id.btn_number_8:
                text += "8";
                break;
            case R.id.btn_number_9:
                text += "9";
                break;
            case R.id.btn_time_symbol:
                text = calculateSymbol(text, 'x');
                break;
            case R.id.btn_divide_symbol:
                text = calculateSymbol(text, '/');
                break;
            case R.id.btn_minus_symbol:
                text = calculateSymbol(text, '-');
                break;
            case R.id.btn_plus_symbol:
                text = calculateSymbol(text, '+');
                break;
            case R.id.btn_c_symbol:
                clear();
                break;
            case R.id.btn_backspace:
                text = backspace(text);
                break;
            case R.id.btn_dot:
                text = dot(text);
                break;
        }

        // jika hasil infinity
        if (tvCalculateAndResult.getText().toString().equals("Infinity")) {
            clear();
        }

        tvCalculateAndResult.setText(text);
    }

    private void clear() {
        this.symbols = "";
        this.text = "";
        tvHistory.setText("");
    }

    private String dot(String text) {
        int length = text.length();
        char strEnd = text.charAt(length - 1);

        if (strEnd == 'x' || strEnd == '/' || strEnd == '+' || strEnd == '-') {
            text += "0.";
        } else {
            text += ".";
        }

        return text;
    }

    /***
     * berfungsi untuk memotong simbol yang double dan simbol yang bertumpuk pada perhitungan
     */
    private String calculateSymbol(String text, char x) {
        if (!TextUtils.isEmpty(tvCalculateAndResult.getText())) {
            int length = text.length();
            char strEnd = text.charAt(length - 1);

            if (strEnd == 'x' || strEnd == '/' || strEnd == '+' || strEnd == '-') {
                text = text.substring(0, (length - 1));
            }

            // collect symbol for calculate
            symbols += x;

            return text + x;
        }
        return "";
    }

    private String backspace(String text) {
        if (TextUtils.isEmpty(tvCalculateAndResult.getText())) {
            text = "";
        } else {
            int length = text.length();
            text = text.substring(0, (length - 1));
        }

        return text;
    }

    private void calculate(String text) {
        if (!TextUtils.isEmpty(tvCalculateAndResult.getText())) {
            String[] textArray = text.split("[x/+-]");
            int textArrayLength = textArray.length;
            int symbolLength = symbols.length();
            double[] numbers = new double[textArrayLength];
            double result = 0;

            // parsing
            for (int i = 0; i < textArrayLength; i++) {
                numbers[i] = Double.parseDouble(textArray[i]);
            }

            for (int i = 0; i < symbolLength; i++) {
                if (symbols.charAt(i) == '/') {
                    result = numbers[i] / numbers[i + 1];
                    numbers[i + 1] = result;
                }
                if (symbols.charAt(i) == 'x') {
                    result = numbers[i] * numbers[i + 1];
                    numbers[i + 1] = result;
                }
                if (symbols.charAt(i) == '+') {
                    result = numbers[i] + numbers[i + 1];
                    numbers[i + 1] = result;
                }
                if (symbols.charAt(i) == '-') {
                    result = numbers[i] - numbers[i + 1];
                    numbers[i + 1] = result;
                }
            }

            String strResult;
            // cari .0 angka dibelakang, jika tidak hanya akan memparsing saja
            if (Double.toString(result).matches(".*[.0]$")) {
                strResult = Double.toString(result).replace(".0", "");
            } else {
                strResult = String.valueOf(result);
            }

            // create tv history
            tvHistory.setText(text);

            // clear all symbol and text change to result
            this.symbols = "";
            this.text = strResult;
            tvCalculateAndResult.setText(strResult);
        }
    }
}

File: Kalkulator/app/src/main/java/com/zukron/kalkulator/MainActivity.kt
package com.zukron.kalkulator

import android.os.Bundle
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*
import androidx.appcompat.app.AppCompatActivity

/**
 * Project name is Kalkulator
 * Created by Zukron Alviandy R on 9/2/2020
 * Contact me if any issues on zukronalviandy@gmail.com
 */
class MainActivity : AppCompatActivity(), View.OnClickListener {
    private var symbols: String = ""
    private var text: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        btn_number_0.setOnClickListener(this)
        btn_number_1.setOnClickListener(this)
        btn_number_2.setOnClickListener(this)
        btn_number_3.setOnClickListener(this)
        btn_number_4.setOnClickListener(this)
        btn_number_5.setOnClickListener(this)
        btn_number_6.setOnClickListener(this)
        btn_number_7.setOnClickListener(this)
        btn_number_8.setOnClickListener(this)
        btn_number_9.setOnClickListener(this)

        btn_c_symbol.setOnClickListener(this)
        btn_time_symbol.setOnClickListener(this)
        btn_divide_symbol.setOnClickListener(this)
        btn_minus_symbol.setOnClickListener(this)
        btn_plus_symbol.setOnClickListener(this)

        btn_backspace.setOnClickListener(this)
        btn_dot.setOnClickListener(this)
        btn_equal_symbol.setOnClickListener(this)
    }

    override fun onClick(view: View?) {
        when (view?.id) {
            R.id.btn_number_0 -> text += "0"
            R.id.btn_number_1 -> text += "1"
            R.id.btn_number_2 -> text += "2"
            R.id.btn_number_3 -> text += "3"
            R.id.btn_number_4 -> text += "4"
            R.id.btn_number_5 -> text += "5"
            R.id.btn_number_6 -> text += "6"
            R.id.btn_number_7 -> text += "7"
            R.id.btn_number_8 -> text += "8"
            R.id.btn_number_9 -> text += "9"
            R.id.btn_time_symbol -> text = calculateSymbol(text, 'x')
            R.id.btn_divide_symbol -> text = calculateSymbol(text, '/')
            R.id.btn_minus_symbol -> text = calculateSymbol(text, '-')
            R.id.btn_plus_symbol -> text = calculateSymbol(text, '+')
            R.id.btn_c_symbol -> clear()
            R.id.btn_backspace -> text = backspace(text)
            R.id.btn_dot -> text = dot(text)
            R.id.btn_equal_symbol -> calculate(tv_calculate_and_result.text.toString())
        }

        // if result is infinity
        if (tv_calculate_and_result.text == "Infinity") {
            clear()
        }

        tv_calculate_and_result.text = text
    }

    /***
     * berfungsi untuk memotong simbol yang double dan simbol yang bertumpuk pada perhitungan
     */
    private fun calculateSymbol(text: String, symbol: Char): String {
        if (!tv_calculate_and_result.text.isBlank()) {
            val length = text.length
            val charEnd = text[length - 1]

            // collect symbol to calculate
            symbols += symbol

            if (charEnd == 'x' || charEnd == '/' || charEnd == '+' || charEnd == '-') {
                return text.substring(0, (length - 1)) + symbol
            }

            return text + symbol
        }
        return ""
    }

    private fun calculate(text: String) {
        if (!text.isBlank()) {
            val textArray = text.split(Regex("[x/+-]"))
            val symbolLength = symbols.length - 1
            val numbers: MutableList<Double> = textArray.map {
                it.toDouble()
            } as MutableList<Double>
            var result: Double? = null

            for (i in 0..symbolLength) {
                if (symbols[i] == '/') {
                    result = numbers[i] / numbers[i + 1]
                    numbers[i + 1] = result
                }

                if (symbols[i] == 'x') {
                    result = numbers[i] * numbers[i + 1]
                    numbers[i + 1] = result
                }

                if (symbols[i] == '+') {
                    result = numbers[i] + numbers[i + 1]
                    numbers[i + 1] = result
                }

                if (symbols[i] == '-') {
                    result = numbers[i] - numbers[i + 1]
                    numbers[i + 1] = result
                }
            }

            // cari .0 angka dibelakang, jika tidak hanya akan memparsing saja
            val strResult = if (result.toString().matches(Regex(".*[.0]$"))) {
                result.toString().replace(".0", "")
            } else {
                result.toString()
            }

            // create history
            tv_history.text = text

            // clear all symbol and text change to result
            this.symbols = ""
            this.text = strResult
            tv_calculate_and_result.text = strResult
        }
    }

    private fun clear() {
        symbols = ""
        text = ""
        tv_history.text = ""
    }

    private fun dot(text: String): String {
        val length = text.length
        val charEnd: Char = text[length - 1]

        return if (charEnd == 'x' || charEnd == '/' || charEnd == '+' || charEnd == '-') {
            text + "0."
        } else {
            "$text."
        }
    }

    private fun backspace(text: String): String {
        return if (text.isBlank()) {
            ""
        } else {
            val length = text.length
            text.substring(0, (length - 1))
        }
    }
}